### ISSUE TITLE
Heatmap: migrate Prometheus heatmap response processing to backend

### ISSUE THREAD

**Issue by leeoniya** [grafana/grafana#46608]
While working on HeatmapNG (https://github.com/grafana/grafana/issues/45495), I looked into what kind of response structure was being produced by a Prometheus heatmap query and what was actually being used to render the old Heatmap. In summary:

Raw response format has a frame-per-bucket structure:

```
▼ {results: {B: {,…}}}
  ▼ results: {B: {,…}}
    ▼ B: {,…}
      ▼ frames: [{schema: {name: "+Inf", refId: "B", meta: {custom: {resultType: "matrix"},…},…}, data: {,…}},…]
          0: {schema: {name: "+Inf", refId: "B", meta: {custom: {resultType: "matrix"},…},…}, data: {,…}}
          1: {schema: {name: "1.0", refId: "B", meta: {custom: {resultType: "matrix"},…},…}, data: {,…}}
          2: {schema: {name: "10.0", refId: "B", meta: {custom: {resultType: "matrix"},…},…}, data: {,…}}
          3: {schema: {name: "2.0", refId: "B", meta: {custom: {resultType: "matrix"},…},…}, data: {,…}}
          4: {schema: {name: "20.0", refId: "B", meta: {custom: {resultType: "matrix"},…},…}, data: {,…}}
          5: {schema: {name: "40.0", refId: "B", meta: {custom: {resultType: "matrix"},…},…}, data: {,…}}
          6: {schema: {name: "5.0", refId: "B", meta: {custom: {resultType: "matrix"},…},…}, data: {,…}}
          7: {schema: {name: "7.0", refId: "B", meta: {custom: {resultType: "matrix"},…},…}, data: {,…}}
        ▼ 8: {schema: {name: "80.0", refId: "B", meta: {custom: {resultType: "matrix"},…},…}, data: {,…}}
          ▼ data: {,…}
            ▼ values: [[1647373605000, 1647373620000, 1647373635000, 1647373650000, 1647373665000, 1647373680000,…],…]
                0: [1647373605000, 1647373620000, 1647373635000, 1647373650000, 1647373665000, 1647373680000,…]
                1: [532.2771221002589, 534.3146289466584, 540.38004299947, 542.4232471762941, 544.4558242764022,…]
          ▼ schema: {name: "80.0", refId: "B", meta: {custom: {resultType: "matrix"},…},…}
              fields: [{name: "Time", type: "time", typeInfo: {frame: "time.Time"}, config: {interval: 15000}},…]
              meta: {custom: {resultType: "matrix"},…}
              name: "80.0"
              refId: "B"
```

Essentially it's a UNIONED series of Prometheus histogram queries. This format is not great for a few reasons:

1. It's bloated - the time field is duplicated, as are the full frame schemas.
2. The count values are cumulative, e.g. `+Inf` values include all buckets < Infinity
3. The frame order looks to be lexicographical rather than ordered by ascending bucket boundary.

To fix 2 and 3 above, there is a bunch of JS transformation code in the Prometheus datasource plugin to make this more useful for the panel to work with:

1. Sort frames in ascending order by bucket boundary name (numerically)
2. De-accumulate each bucket, so it only contains values in the bucket itself

https://github.com/grafana/grafana/blob/468def0c003482c153f92a1c480b2f98d23a04d6/public/app/plugins/datasource/prometheus/result_transformer.ts#L273-L278

https://github.com/grafana/grafana/blob/468def0c003482c153f92a1c480b2f98d23a04d6/public/app/plugins/datasource/prometheus/result_transformer.ts#L77

i inadvertently re-created this logic for HeatmapNG before realizing it was already being done in the datasource JS and disabled it in a follow-up https://github.com/grafana/grafana/pull/46309.

Really, we'd just like to remove this frontend munging and move this processing to the Go side, so we can expect receive the ready-to-go format in the histogram response. In doing this, we additionally want to switch from [multi-frame] frame-per-bucket to a [single frame] field-per-bucket format, thus eliminating the redundancy of schema and Time field duplication. We can make a simple adjustment to the old Heatmap panel to simply re-create the current frame-per-bucket format until it is decomissioned and replaced with HeatmapNG (very soon).

cc @toddtreece @aocenas

**Comment by leeoniya** [comment]
> In doing this, we additionally want to switch from [multi-frame] frame-per-bucket to a [single frame] field-per-bucket format, thus eliminating the redundancy of schema and Time field duplication. We can make a simple adjustment to the old Heatmap panel to simply re-create the current frame-per-bucket format

my WIP for this: https://github.com/grafana/grafana/pull/46611

**Comment by evandandrea** [comment]
@itsmylife is going to reach out to @leeoniya to build context / understanding next steps on this. https://github.com/grafana/grafana-plugin-sdk-go/pull/507 is also worth digging into.

**Comment by quentinmit** [comment]
Note that this is currently only used for the Prometheus datasource, which means that the transformation can't be reused for other kinds of datasource (even ones that contain Prometheus-format histograms). So I would actually like to see it move the other direction, in a way that it can be applied to the output of any datasource.

**Comment by quentinmit** [comment]
I don't believe there is any support for handling histograms in a datasource-agnostic way yet.
